lib/idb.mjs

```diff
@@ -0,0 +1,304 @@
+  return new ObjectStore(this._tx.objectStore.apply(this._tx, arguments));
+};
+
+proxyProperties(Transaction, '_tx', [
+  'objectStoreNames',
+  'mode'
+]);
+
+proxyMethods(Transaction, '_tx', IDBTransaction, [
+  'abort'
+]);
+
+function UpgradeDB(db, oldVersion, transaction) {
+  this._db = db;
+  this.oldVersion = oldVersion;
+  this.transaction = new Transaction(transaction);
+}
+
+UpgradeDB.prototype.createObjectStore = function() {
+  return new ObjectStore(this._db.createObjectStore.apply(this._db, arguments));
+};
+
+proxyProperties(UpgradeDB, '_db', [
+  'name',
+  'version',
+  'objectStoreNames'
+]);
+
+proxyMethods(UpgradeDB, '_db', IDBDatabase, [
+  'deleteObjectStore',
+  'close'
+]);
+
+function DB(db) {
+  this._db = db;
+}
+
+DB.prototype.transaction = function() {
+  return new Transaction(this._db.transaction.apply(this._db, arguments));
+};
+
+proxyProperties(DB, '_db', [
+  'name',
+  'version',
+  'objectStoreNames'
+]);
+
+proxyMethods(DB, '_db', IDBDatabase, [
+  'close'
+]);
+
+// Add cursor iterators
+// TODO: remove this once browsers do the right thing with promises
+['openCursor', 'openKeyCursor'].forEach(function(funcName) {
+  [ObjectStore, Index].forEach(function(Constructor) {
+    // Don't create iterateKeyCursor if openKeyCursor doesn't exist.
+    if (!(funcName in Constructor.prototype)) return;
+
+    Constructor.prototype[funcName.replace('open', 'iterate')] = function() {
+      var args = toArray(arguments);
+      var callback = args[args.length - 1];
+      var nativeObject = this._store || this._index;
+      var request = nativeObject[funcName].apply(nativeObject, args.slice(0, -1));
+      request.onsuccess = function() {
+        callback(request.result);
+      };
+    };
+  });
+});
+
+// polyfill getAll
+[Index, ObjectStore].forEach(function(Constructor) {
+  if (Constructor.prototype.getAll) return;
+  Constructor.prototype.getAll = function(query, count) {
+    var instance = this;
+    var items = [];
+
+    return new Promise(function(resolve) {
+      instance.iterateCursor(query, function(cursor) {
+        if (!cursor) {
+          resolve(items);
+          return;
+        }
+        items.push(cursor.value);
+
+        if (count !== undefined && items.length == count) {
+          resolve(items);
+          return;
+        }
+        cursor.continue();
+      });
+    });
+  };
+});
+
+export default {
+  open: function(name, version, upgradeCallback) {
+    var p = promisifyRequestCall(indexedDB, 'open', [name, version]);
+    var request = p.request;
+
+    if (request) {
+      request.onupgradeneeded = function(event) {
+        if (upgradeCallback) {
+          upgradeCallback(new UpgradeDB(request.result, event.oldVersion, request.transaction));
+        }
+      };
+    }
+
+    return p.then(function(db) {
+      return new DB(db);
+    });
+  },
+  delete: function(name) {
+    return promisifyRequestCall(indexedDB, 'deleteDatabase', [name]);
+  }
```

Comment:
It feels like it would be more module-like to export open & delete separately. Is this something you thought about?

Reply:
Agreed, that's better. I've changed to separate exports in fd0ca27dbcaec38e8ac1acdf9805aa3d99cbb8b2. Note that I had to name the function `_delete` to avoid:
```
Error: Unexpected keyword 'delete'
```
but it is still exported as `delete`